### PR TITLE
Action bot

### PR DESCRIPTION
**Pull request by goldenratio1618**

### CODE REVIEW

File: scripts/train.py [hunk 1]
```diff
@@ -87,10 +89,10 @@
 def build_agent():
     logger.info("Creating generator...")
     if TRAIN_GENERATOR:
-        gen_model = MinionsDiscriminator(d_model=GEN_D_MODEL, depth=GEN_DEPTH)  # TODO - make generator model instead
+        gen_model = MinionsActionBot(d_model=GEN_D_MODEL, depth=GEN_DEPTH)  # TODO - make generator model instead
```
Comment (dfarhi) [inline scripts/train.py:92]: remove TODO

File: scripts/train.py [hunk 2]
```diff
@@ -87,10 +89,10 @@
...
         logger.info("Generator model initialized:")
         logger.info(gen_model)
-        gen_translator = Translator()  # TODO - make gen translator
+        gen_translator = Translator("generator")  # TODO - make gen translator
```
Comment (dfarhi) [inline scripts/train.py:95]: remove TODO

File: minionsai/discriminator_only/translator.py [hunk 3]
```diff
@@ -57,30 +85,122 @@ def translate(self, game: Game):
...
-            'remaining_turns': obs['remaining_turns'],
-            'money': obs['money'],
-            'opp_money': obs['opp_money'],
-            'score_diff': obs['score_diff']
-        } for b in all_boards]
\ No newline at end of file
+
+        new_obs = []
+        for b in all_boards:
+            d = OrderedDict()
```
Comment (dfarhi) [inline minionsai/discriminator_only/translator.py:197]: It would be better here to reuse the ordering in `obs`, rather than rewriting it - now there are two sources of truth about the correct order in the codebase.

File: minionsai/discriminator_only/translator.py [hunk 4]
```diff
@@ -57,30 +85,122 @@ def translate(self, game: Game):
...
+            d['board'] = b # shape is [batch, num_items=BOARD_SIZE^2]
+            d['money'] = obs['money']
```
Comment (dfarhi) [inline minionsai/discriminator_only/translator.py:199]: Can't you just do:
```
for key, value in obs.items():
   if key == 'board':
       d[key] = b
    else:
        d[key] = v
```
That way it won't need the hardcoded extra handling for phase either.

File: minionsai/discriminator_only/translator.py [hunk 5]
```diff
@@ -57,30 +85,122 @@ def translate(self, game: Game):
...
+            d['phase'] = np.array([[phase]])
+            # legal_actions = [self.valid_actions(game).flatten()]
+            # d['legal_actions'] = np.array([])
+        return d
+
+
+
+    def get_possible_legal_actions(self):
+        possibly_legal = np.zeros((self.num_things, self.num_things), bool)
+        for i in range(BOARD_SIZE ** 2 + 1):
```
Comment (dfarhi) [inline minionsai/discriminator_only/translator.py:104]: This loop going to `BOARD_SIZE ** 2 +1` only works if `self.spawn_pos = BOARD_SIZE` I think. It might be simpler to just have the main loop over `BOARD_SIZE ** 2`, and handle spawn separately.

As an example of why it's confusing with the +1 in the loop: if I understand right the entire last loop over `j` is guaranteed to do nothing.

File: minionsai/discriminator_only/translator.py [hunk 6]
```diff
@@ -57,30 +85,122 @@ def translate(self, game: Game):
...
+                # money -> board = spawn
+                # money -> money = advance phase
+                if i == self.spawn_pos:
+                    possibly_legal[i,j] = True
+                else:
+                    if j == self.spawn_pos:
+                        continue
+                    x1, y1 = convert_loc_to_pos(i)
+                    x2, y2 = convert_loc_to_pos(j)
+                    if abs(x1-x2) <= MAX_SPEED_OR_RANGE and abs(y1-y2) <= MAX_SPEED_OR_RANGE:
```
Comment (dfarhi) [inline minionsai/discriminator_only/translator.py:115]: I think we have a nice distance-calculating utility in `engine.py` or `game_util.py` or somewhere, that you could use to get even more precise measurement of distance. But probably not a big deal, fine to leave it as is.

File: minionsai/discriminator_only/translator.py [hunk 7]
```diff
@@ -57,30 +85,122 @@ def translate(self, game: Game):
...
+                        x2, y2 = convert_loc_to_pos(j)
+                        act = SpawnAction(ZOMBIE, (x2,y2))
+                        if game.process_single_spawn(act, False)[0]:
+                            legal[i,j] = True
+                    else:
+                        # wrong phase, action is illegal
+                        if game.phase != Phase.MOVE:
+                            continue
+                        # move action
+                        x1, y1 = convert_loc_to_pos(i)
```
Comment (dfarhi) [inline minionsai/discriminator_only/translator.py:142]: nitpick: can you avoid code duplication by using:
```
act = self.untranslate_action((i, j)
```
instead of duplicating the code here?

In fact would something go wrong if this whole for loop were just:
```
act = self.untranslate_action(i, j)
result, _ = game.process_single_action(act, False)
legal[i, j] = result
```
?

File: minionsai/discriminator_only/translator.py [hunk 8]
```diff
@@ -57,30 +85,122 @@ def translate(self, game: Game):
...
+                    act = MoveAction((x1,y1), (x2,y2))
+                    actions[i][j] = act
+        return actions
+
+    def untranslate_action(self, action):
+        untranslated_action = self.untranslated_actions[action[0]][action[1]]
+        assert not (untranslated_action is None)
+        return untranslated_action
+
+    # def process_action(self, game, action):
```
Comment (dfarhi) [inline minionsai/discriminator_only/translator.py:175]: nitpick: remove commented code?

File: minionsai/discriminator_only/translator.py [hunk 9]
```diff
@@ -57,30 +85,122 @@ def translate(self, game: Game):
...
+                else:
+                    if j == self.spawn_pos:
+                        continue
+                    x1, y1 = convert_loc_to_pos(i)
+                    x2, y2 = convert_loc_to_pos(j)
+                    if abs(x1-x2) <= MAX_SPEED_OR_RANGE and abs(y1-y2) <= MAX_SPEED_OR_RANGE:
+                        possibly_legal[i,j] = True
+        return possibly_legal
+
+    def valid_actions(self, game: Game):
```
Comment (dfarhi) [inline minionsai/discriminator_only/translator.py:119]: It might be good to unit test these valid-action functions. I imagine if they have subtle edge-case errors we'll get horribly confused, and they're pretty self contained. e.g. you could assemble a random game state and just test that the valid_actions array is exactly what it should be.

File: minionsai/discriminator_only/translator.py [hunk 10]
```diff
@@ -1,11 +1,21 @@
 
 
 from functools import lru_cache
+from operator import truediv
+from tkinter.tix import MAX
```
Comment (dfarhi) [inline minionsai/discriminator_only/translator.py:5]: What is tkinter.tix? Isn't that some plotting library? I suspect your editor added this without your permission and it's unused. Also the xmlrpc thing below.

**Review by dfarhi — COMMENTED** [review]
Looks amazing! I left a zillion small comments which I think will make it more future-proof, but none of them are really necessary.

File: minionsai/action.py [hunk 11]
```diff
@@ -51,5 +62,12 @@ def __repr__(self):
...
     @staticmethod
     def from_single_list(actions):
+        move_actions = []
+        spawn_actions = []
+        for action in actions:
+            if type(action) is MoveAction:
+                move_actions.append(action)
+            elif type(action) is SpawnAction:
+                spawn_actions.append(action)
         # TODO return ActionList from a list of single actions
```
[the remark this answers is not in the document]
Comment (goldenratio1618) [reply minionsai/action.py:72]: I think we'd probably need to refactor a good amount of code for that, I think a good amount of code (eg: HumanAIAgent) doesn't currently use the AdvancePhaseActions. Could be worth doing to get all the code to use the same framework.

File: minionsai/discriminator_only/translator.py (hunk 7 above)
Comment (goldenratio1618) [reply minionsai/discriminator_only/translator.py:142]: the game engine handles trying to move during the wrong phase differently than it handles trying to take an illegal action during a given phase, for some reason. In the former case, the code will crash, while in the latter case, it will return `False` and a status code. I simplified the code a bit, but it'll require a nontrivial refactor of the game engine (and things that relied on the game engine throwing errors instead of getting into weird, unreachable states) to make that work
